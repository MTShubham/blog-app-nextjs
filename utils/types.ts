export type CategoryType = {
    title: string, 
    description: string
}

export type ImageType = {
    url: string,
    alt: string
}

export type AuthorType = {
    profileImage: ImageType,
    name: string,
    organization: string
}

export type PostType = {
    slug: string,
    title: string,
    subtitle: string,
    posterImage: ImageType,
    categories: CategoryType[],
    body: [],
    _id: string,
    author: AuthorType,
    _createdAt: string,
    _updatedAt: string,
}