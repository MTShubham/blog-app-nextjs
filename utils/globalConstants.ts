export const MAX_BOOKMARK_ALLOWED = 5
export const REVALIDATE_TIME = 5