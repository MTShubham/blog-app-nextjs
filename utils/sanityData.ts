import sanityClient from "@/sanity/sanity.client";
import { groq } from "next-sanity";

export const getAllPosts = async () => {
  const posts = await sanityClient.fetch((groq`
    *[_type=="post"] | order(_createdAt desc) {
        ...,
        "slug": slug.current,
        "posterImage": {"alt": posterImage.alt, "url": posterImage.asset->.url},
        "author": author->{
          ...,
          "profileImage": {"alt": profileImage.alt, "url":profileImage.asset->.url}
        },
        "categories": categories[]->
      }
  `), { cache: 'no-store' });
  return posts;
}

export const getPostsByBookmarkedPostIds = async (bookmarkedPostIds: string[]) => {
  const filteredPosts = await sanityClient.fetch(groq`
        *[_type=="post" && _id in $bookmarkedPostIds] {
        ...,
        "slug": slug.current,
        "posterImage": {"alt": posterImage.alt, "url": posterImage.asset->.url},
        "author": author->{
        ...,
        "profileImage": {"alt": profileImage.alt, "url":profileImage.asset->.url}
        },
        "categories": categories[]->
        }
        `, {
    bookmarkedPostIds
  });
  return filteredPosts;
}