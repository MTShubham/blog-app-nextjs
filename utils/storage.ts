export function setLocalStorage(key: string, value: any) {
    localStorage.setItem(key, JSON.stringify(value));
    window.dispatchEvent(new Event("storage"));
}

export function getLocalStorage(key: string) {
    let value = localStorage.getItem(key);
    if (value) {
        JSON.parse(value);
    }
    return value ? value : '';
}

export function removeLocalStorage(key: string) {
    localStorage.removeItem(key);
    window.dispatchEvent(new Event("storage"));
}