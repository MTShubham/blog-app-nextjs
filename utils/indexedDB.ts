import { openDB } from 'idb';
import { MAX_BOOKMARK_ALLOWED } from './globalConstants';

type User = {
    username: string,
    password: string
}

export async function initIndexedDB() {
    const db = await openDB('blogs', 2, {
        upgrade(db) {
            if (!db.objectStoreNames.contains('users')) {
                db.createObjectStore('users', { keyPath: 'username' });
            }
            if (!db.objectStoreNames.contains('bookmarks')) {
                db.createObjectStore('bookmarks', { keyPath: 'username' });
            }
        },
    });
    return db;
}

export async function saveUser(db, user: User) {
    const transaction = await db.transaction('users', 'readwrite');
    let users = transaction.objectStore('users');
    const request = await users.add(user);
    if (request) {
        return { success: true, status: 201, msg: 'Created an user account' };
    }
    else {
        return { success: false, status: 500, msg: 'Error in creating user' };
    }
}

export async function loginUser(db, user: User) {
    const transaction = await db.transaction('users', 'readonly');
    let users = transaction.objectStore('users');
    const foundUser = await users.get(user.username);
    if (foundUser?.password === user.password && foundUser !== undefined) {
        return { success: true, status: 200, msg: 'Successful login' };
    }
    else {
        return { success: false, status: 500, msg: 'Invalid credentials' };
    }
}

export async function getBookmarkedPosts(db, username: string) {
    const transaction = await db.transaction('bookmarks', 'readwrite');
    let bookmarks = transaction.objectStore('bookmarks');
    let userBookmarks = await bookmarks.get(username);
    let postIds = userBookmarks;
    if (userBookmarks) {
        ({ postIds } = userBookmarks);
        return { success: true, status: 200, postIds };
    }
    return { success: false, status: 500, msg: 'Unable to fetch bookmarked posts' };
}

export async function removePostBookmark(db, username: string, postId: string) {
    const transaction = await db.transaction('bookmarks', 'readwrite');
    let bookmarks = transaction.objectStore('bookmarks');
    let userBookmarks = await bookmarks.get(username);
    let { postIds } = userBookmarks;
    let postIdsAfterRemoval = postIds.filter((id: string) => postId !== id)
    const request = await bookmarks.put({ username, postIds: postIdsAfterRemoval });
    if (request) {
        return { success: true, status: 201, postIds: postIdsAfterRemoval };
    }
    else {
        return { success: false, status: 500, msg: 'Unable to remove bookmark' };
    }
}

export async function bookmarkPost(db, username: string, postId: string) {
    const transaction = await db.transaction('bookmarks', 'readwrite');
    let bookmarks = transaction.objectStore('bookmarks');
    let userBookmarks = await bookmarks.get(username);
    let postIds;
    if (userBookmarks) {
        ({ postIds } = userBookmarks);
    }
    else {
        postIds = [];
    }
    if (postIds.length === MAX_BOOKMARK_ALLOWED) {
        return { success: false, status: 400, msg: 'Only 5 bookmarks are allowed per user and you have achieved that. So you cannot bookmark more blog posts' };
    }
    postIds.push(postId);
    const request = await bookmarks.put({ username, postIds: postIds });
    if (request) {
        return { success: true, status: 200, postIds };
    }
    else {
        return { success: false, status: 500, msg: 'Unable to bookmark due to server error' };
    }
}   