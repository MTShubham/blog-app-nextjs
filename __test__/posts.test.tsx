import localStorageMock from '@/__mock__/localStorage';
import { getAllPosts } from '@/utils/sanityData';

const getPostsData = async () => {
    return await getAllPosts();
}

describe('Index page', () => {
    
    test('Getting all posts', async () => {
        let data = await getPostsData();
        expect(data).not.toHaveLength(0);
    });

    // test('Title of all the posts loaded or not', async () => {
    //     let data = await getPostsData();
    //     data.forEach((post) => {
    //         expect(post.title).not.toBeNull();
    //     });
    // })

    // test('Author of all the posts loaded or not', async () => {
    //     let data = await getPostsData();
    //     data.forEach((post) => {
    //         expect(post.author).not.toBeNull();
    //     });
    // })

    // test('Whether all the poster images of the post loaded or not', async () => {
    //     let data = await getPostsData();
    //     data.forEach((post) => {
    //         expect(post.posterImage.alt).not.toBeNull();
    //         expect(post.posterImage.url).not.toBeNull();
    //     });
    // })

    // test('Body of all the posts loaded or not', async () => {
    //     let data = await getPostsData();
    //     data.forEach((post) => {
    //         expect(post.body).not.toBeUndefined();
    //     });
    // })
});