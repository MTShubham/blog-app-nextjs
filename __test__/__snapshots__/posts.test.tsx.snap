// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page Whether all the poster images of the post loaded or not 1`] = `
{
  "alt": "Continuous learning",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 2`] = `
{
  "alt": "Self believe",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 3`] = `
{
  "alt": "Audio to software",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 4`] = `
{
  "alt": "Stress and happiness",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 5`] = `
{
  "alt": "AI future",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 6`] = `
{
  "alt": "Strategies for success",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 7`] = `
{
  "alt": "Ink",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 8`] = `
{
  "alt": "Healthcare",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 9`] = `
{
  "alt": "Website Terms and Conditions",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 10`] = `
{
  "alt": "ChatGPT Developers",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 11`] = `
{
  "alt": "Technical",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 12`] = `
{
  "alt": "Resilience",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 13`] = `
{
  "alt": "Software Developer",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 14`] = `
{
  "alt": "Learning from mistakes",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 15`] = `
{
  "alt": "Relaxing Life",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 16`] = `
{
  "alt": "Life Wheel",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 17`] = `
{
  "alt": "rise of PhonePe",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 18`] = `
{
  "alt": "Data Science",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 19`] = `
{
  "alt": "Tech and ML",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 20`] = `
{
  "alt": "Postr Image",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 21`] = `
{
  "alt": "UX images",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 22`] = `
{
  "alt": "Software Developer",
  "url": Any<String>,
}
`;

exports[`Index page Whether all the poster images of the post loaded or not 23`] = `
{
  "alt": "react structure",
  "url": Any<String>,
}
`;
